pkg/controllers/cluster/actions/upgrade_version.go

```diff
@@ -3,54 +3,966 @@ package actions
+				return ActionFailure, errors.Wrap(err, "set upgrade strategy")
+			}
+		}
+	}
+
+	return ActionSuccess, nil
+}
+
+func (a *ClusterVersionUpgrade) fsm() *fsm.StateMachine {
+	state := BeginUpgrade
+	if a.Cluster.Status.Upgrade != nil {
+		state = fsm.StateType(a.Cluster.Status.Upgrade.State)
+	}
+
+	return &fsm.StateMachine{
+		TransitionHook: a.onStateTransition,
+		Current:        state,
+		States: map[fsm.StateType]fsm.State{
+			BeginUpgrade: {
+				Action: a.beginUpgrade,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: CheckSchemaAgreement,
+				}},
+			CheckSchemaAgreement: {
+				Action: a.checkSchemaAgreement,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: CreateSystemBackup,
+				}},
+			CreateSystemBackup: {
+				Action: a.createSystemBackup,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: FindNextRack,
+				}},
+			FindNextRack: {
+				Action: a.findNextRack,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess:    UpgradeImageInPodSpec,
+					AllRacksUpgraded: ClearSystemBackup,
+				},
+			},
+			UpgradeImageInPodSpec: {
+				Action: a.updateRackSpec,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: FindNextNode,
+				},
+			},
+			RestoreUpgradeStrategy: {
+				Action: a.restoreUpgradeStrategy,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: FindNextRack,
+				},
+			},
+			FindNextNode: {
+				Action: a.findNextNode,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess:    EnableMaintenanceMode,
+					AllNodesUpgraded: RestoreUpgradeStrategy,
+				},
+			},
+			EnableMaintenanceMode: {
+				Action: a.enableMaintenanceMode,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: DrainNode,
+				},
+			},
+			DrainNode: {
+				Action: a.drainNode,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: BackupData,
+				},
+			},
+			BackupData: {
+				Action: a.createDataBackup,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: DisableMaintenanceMode,
+				},
+			},
+			DisableMaintenanceMode: {
+				Action: a.disableMaintenanceMode,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: DeletePod,
+				},
+			},
+			DeletePod: {
+				Action: a.deletePod,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: ValidateUpgrade,
+				},
+			},
+			ValidateUpgrade: {
+				Action: a.validateUpgrade,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: ClearDataBackup,
+				},
+			},
+			ClearDataBackup: {
+				Action: a.clearDataBackup,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: FindNextNode,
+				},
+			},
+			ClearSystemBackup: {
+				Action: a.clearSystemBackup,
+				Events: map[fsm.EventType]fsm.StateType{
+					ActionSuccess: FinishUpgrade,
+				},
+			},
+			FinishUpgrade: {
+				Action: a.finishUpgrade,
+			},
+		},
+	}
+}
+
+func contains(arr []string, v string) bool {
```

Comment:
This little sh** deserves it's own package...

Reply:
Totally agree, I'm surprised there's no such function in `strings` package. I guess they are waiting for generics.